src/node_report.cc

```diff
@@ -0,0 +1,808 @@
+  fprintf(fp, "================================================================================\n");
+  fprintf(fp, "==== NodeReport ================================================================\n");
+  fprintf(fp, "\nEvent: %s, location: \"%s\"\n", message, location);
+  fprintf(fp, "Filename: %s\n", filename);
+
+  // Print dump event and module load date/time stamps
+#ifdef _WIN32
+  fprintf(fp, "Dump event time:  %4d/%02d/%02d %02d:%02d:%02d\n",
+          tm_struct.wYear, tm_struct.wMonth, tm_struct.wDay,
+          tm_struct.wHour, tm_struct.wMinute, tm_struct.wSecond);
+  fprintf(fp, "Module load time: %4d/%02d/%02d %02d:%02d:%02d\n",
+          loadtime_tm_struct.wYear, loadtime_tm_struct.wMonth, loadtime_tm_struct.wDay,
+          loadtime_tm_struct.wHour, loadtime_tm_struct.wMinute, loadtime_tm_struct.wSecond);
+#else  // UNIX, OSX
+  fprintf(fp, "Dump event time:  %4d/%02d/%02d %02d:%02d:%02d\n",
+          tm_struct.tm_year+1900, tm_struct.tm_mon+1, tm_struct.tm_mday,
+          tm_struct.tm_hour, tm_struct.tm_min, tm_struct.tm_sec);
+  fprintf(fp, "Module load time: %4d/%02d/%02d %02d:%02d:%02d\n",
+          loadtime_tm_struct.tm_year+1900, loadtime_tm_struct.tm_mon+1, loadtime_tm_struct.tm_mday,
+          loadtime_tm_struct.tm_hour, loadtime_tm_struct.tm_min, loadtime_tm_struct.tm_sec);
+#endif
+  // Print native process ID
+  fprintf(fp, "Process ID: %d\n", pid);
+  fflush(fp);
+
+  // Print Node.js and OS version information
+  PrintVersionInformation(fp);
+  fflush(fp);
+
+// Print summary JavaScript stack backtrace
+  PrintJavaScriptStack(fp, isolate, event, location);
+  fflush(fp);
+
+  // Print native stack backtrace
+  PrintNativeStack(fp);
+  fflush(fp);
+
+  // Print V8 Heap and Garbage Collector information
+  PrintGCStatistics(fp, isolate);
+  fflush(fp);
+
+  // Print OS and current thread resource usage
+#ifndef _WIN32
+  PrintResourceUsage(fp);
+  fflush(fp);
+#endif
+
+  // Print libuv handle summary (TODO: investigate failure on Windows)
+  // Note: documentation of the uv_print_all_handles() API says "This function
+  // is meant for ad hoc debugging, there is no API/ABI stability guarantee"
+  // http://docs.libuv.org/en/v1.x/misc.html
+#ifndef _WIN32
+  fprintf(fp, "\n================================================================================");
+  fprintf(fp, "\n==== Node.js libuv Handle Summary ==============================================\n");
+  fprintf(fp,"\n(Flags: R=Ref, A=Active, I=Internal)\n");
+  fprintf(fp,"\nFlags Type     Address\n");
+  uv_print_all_handles(NULL, fp);
+  fflush(fp);
+#endif
+
+  // Print operating system information
+  PrintSystemInformation(fp, isolate);
+
+  fprintf(fp, "\n================================================================================\n");
+  fflush(fp);
+  fclose(fp);
+
+  fprintf(stderr, "Node.js report completed\n");
+  if (name != NULL) {
+    snprintf(name, NR_MAXNAME + 1, "%s", filename);  // return the NodeReport file name
```

Comment:
Declare `name` like this in the function's parameter list: `char (*name)[NR_MAXNAME + 1]`.  That way the caller cannot pass in a wrong-sized buffer.